numpy/core/numeric.py

```diff
@@ -2379,6 +2379,9 @@ def within_tol(x, y, atol, rtol):
 def _array_equal_dispatcher(a1, a2, equal_nan=None):
     return (a1, a2)
 
+def _dtype_cannot_hold_nan(dtype):
+    return dtype in {np.int8, np.int16, np.int32, np.int64}
```

Comment:
```suggestion
_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan(dtype):
    return dtype in _no_nan_dtypes
```
This is much faster, a benchmark:
```
import numpy as np

def _dtype_cannot_hold_nan(dtype):
    return dtype in {np.int8, np.int16, np.int32, np.int64}

_no_nan_dtypes = {np.int8, np.int16, np.int32, np.int64}

def _dtype_cannot_hold_nan_2(dtype):
    return dtype in _no_nan_dtypes

dtype=np.array([1.]).dtype

%timeit _dtype_cannot_hold_nan(dtype)
%timeit _dtype_cannot_hold_nan_2(dtype)
%timeit dtype in _no_nan_dtypes
```
results in
```
507 ns ± 63.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
112 ns ± 1.94 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
57.9 ns ± 5.47 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```
In inline version is the fastest, so that would be an option as well.

Reply:
One question is do we want to let people add things to `_no_nan_dtypes` by monkeypatching.